ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh

```diff
@@ -460,6 +461,120 @@ function check_history() {
     fi
 }
 
+# Upgrade vsphere hardware version
+function vsphere_hw_upgrade() {
+    source "${SHARED_DIR}/govc.sh"
+
+    master_nodes=$(oc get nodes -l node-role.kubernetes.io/master -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}')
+    worker_nodes=$(oc get nodes -l node-role.kubernetes.io/worker -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}')
+
+    first_node=$(echo ${master_nodes} | cut -d" " -f1)
+    cluster_name=$(<"${SHARED_DIR}"/clustername.txt)
+    vm_path="/${GOVC_DATACENTER}/vm/${cluster_name}"
+
+    current_hw_version=$(govc vm.info -json=true ${vm_path}/${first_node} | jq -r .VirtualMachines[0].Config.Version)
+    echo "Current hardware version is: ${current_hw_version}"
+    hw_version=$(echo ${current_hw_version} | cut -d'-' -f2)
+
+    if (( hw_version >= VSPHERE_HW_VERSION )); then
+        echo "VM hardware version is not less than ${VSPHERE_HW_VERSION}, no upgrade is required"
+        exit 0
+    fi
+
+    echo "Updating VM hardware version to ${VSPHERE_HW_VERSION}"
+    # shellcheck disable=SC2068
+    for node in ${master_nodes[@]} ${worker_nodes[@]}; do
+        echo "Marking the node as unschedulable"
+        oc adm cordon ${node}
+        if [[ "${worker_nodes[*]}" =~ ${node} ]]; then
+            echo "Evacuate the pods from the compute node"
+            oc adm drain ${node} --force=true --delete-emptydir-data --ignore-daemonsets
```

Comment:
An active `PodDisruptionBudget` may break and `oc adm drain` indefinitely, having a heavy operation like this does not always guarantee the best results. An ideal state would be that the hardware meets requirements by default. If vSphere has so much added complexity, I suggest running the automation on a different cloud and leaving this to manual operation? 

WDYT @jianlinliu @evakhoni @jinyunma ?

Reply:
I guess we have to hit `drain node` cases in the future even we did no to that this time.

Reply:
agree, there is similar bug https://issues.redhat.com/browse/OCPBUGS-2353 when updating machineconfig, drain node failed due to break the pdb of openshift-cluster-csi-drivers, but bug is closed now and refer to the enhancement https://issues.redhat.com/browse/RFE-1367.

I remembered that the operation of drain node is one step in doc [Updating the virtual hardware for compute nodes on vSphere](https://docs.openshift.com/container-platform/4.11/updating/updating-hardware-on-nodes-running-on-vsphere.html#update-vsphere-virtual-hardware-on-compute-nodes_updating-hardware-on-nodes-running-in-vsphere), the options of drain node referred to [here](https://docs.openshift.com/container-platform/4.11/nodes/nodes/nodes-nodes-working.html#nodes-nodes-working-evacuating_nodes-nodes-working) and based on our testing, so agree that it might not guarantee passed on every time.

Reply:
Another point, I am thinking if it is possible to split the `drain node` operations into a dedicated step, so that we can opt-in the step when it is required.